cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/executor/strategy/AbstractReplicaMovementStrategy.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2018 LinkedIn Corp. Licensed under the BSD 2-Clause License (the "License"). See License in the project root for license information.
+ */
+
+package com.linkedin.kafka.cruisecontrol.executor.strategy;
+
+import com.linkedin.kafka.cruisecontrol.executor.ExecutionProposal;
+import com.linkedin.kafka.cruisecontrol.executor.ExecutionTask;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.SortedSet;
+import java.util.TreeSet;
+import org.apache.kafka.common.Cluster;
+
+
+/**
+ * An abstract class for replica movement strategy. This class will be extended to create custom strategy to determine the
+ * execution order the replica movement tasks.
+ */
+public abstract class AbstractReplicaMovementStrategy implements ReplicaMovementStrategy {
+
+  /**
+   * Generate a comparator for replica movement task which incorporate the strategy to apply. The "smaller" task will have
+   * higher execution priority.
+   *
+   * @param cluster The current cluster state.
+   * @return The comparator of task.
+   */
+  public abstract Comparator<ExecutionTask> taskComparator(Cluster cluster);
+
+  @Override
+  public ReplicaMovementStrategy chain(ReplicaMovementStrategy strategy) {
+    if (!(strategy instanceof AbstractReplicaMovementStrategy)) {
```

Comment:
Would it be possible to move this sanity check to `KafkaCruiseControlConfig`?

Reply:
I think this is a constraint only for `sanityCheckReplicaMovementStrategies`(abstract class) not `ReplicaMovementStrategy`( interface). If we move this check to `KafkaCruiseControlConfig` we actually place an unwanted restriction that Cruise Control can only use extend class of the abstract class instead of the interface and jeopardize the generic property.